WalletWasabi/Crypto/ZeroKnowledge/Transcript.cs

```diff
@@ -0,0 +1,185 @@
+			// FIXME why does this do an infinite loop?
+			// Statement(tag, publicPoint, generators);
+			//
+			// i expected overloading to distinguish params GE[] and Enumerable<GE>?
+			// above definition copypasted here
+
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+
+			// TODO add guard to ensure generators is non empty?
+			foreach (var generator in generators)
+			{
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+			}
+
+			var concatenation = generators.SelectMany(x => x.ToBytes());
+			var msg = ByteHelpers.Combine(BitConverter.GetBytes(tag.Length), tag, BitConverter.GetBytes(generators.Count()), concatenation.ToArray());
+			using var sha256 = System.Security.Cryptography.SHA256.Create();
+			var hash = sha256.ComputeHash(msg);
+
+			state.AssociatedData(Encoding.UTF8.GetBytes("statement"));
+			state.AssociatedData(hash);
+			state.AssociatedData(publicPoint.ToBytes());
+		}
+
+		// generate synthetic nonce using current state combined with additional randomness
+		public Scalar GenerateNonce(Scalar secret, WasabiRandom? random = null) {
+			// to integrate prior inputs for deterministic component of nonce
+			// generation, first clone the state at the current point in the
+			// transcript, which should already have the statement tag and public
+			// inputs committed.
+			var forked = state.Clone();
+
+			// add secret inputs as key material
+			forked.Key(secret.ToBytes());
+
+			// get randomness from system if no random source specified
+			var fromRng = new byte[32];
+			if (random is null )
+			{
+				random = new SecureRandom();
+			}
+			random.GetBytes(fromRng);
+
+			// add additional randomness as associated data
+			forked.AssociatedData(fromRng);
+
+			// generate a new scalar using this updated state as a seed
+			return new Scalar(forked.PRF());
+		}
+
+		public void NonceCommitment(GroupElement nonce) {
+			Guard.False($"{nameof(nonce)}.{nameof(nonce.IsInfinity)}", nonce.IsInfinity);
+			state.AssociatedData(Encoding.UTF8.GetBytes("nonce commitment"));
+			state.AssociatedData(nonce.ToBytes());
+		}
+
+		// generate Fiat Shamir challenges
+		public Scalar GenerateChallenge() {
+			// generate a new scalar using current state as a seed
+			return new Scalar(state.PRF());
+		}
+
+		// implements a stepping stone towards STROBE
+		private class State {
+			private byte[] h;
+
+			public State(byte[] initial) {
+				h = initial;
+			}
+
+			public State Clone() {
+				return new State(h);
+			}
+
+			private void Absorb(StrobeFlags flags, byte[] data) {
+				// modeled after Noise's MixHash operation, in line with reccomendation in
+				// per STROBE paper appendix B, for when not using a Sponge function.
+				// stepping stone towards STROBE with Keccak.
+				using var sha256 = System.Security.Cryptography.SHA256.Create();
+				h = sha256.ComputeHash(ByteHelpers.Combine(h, BitConverter.GetBytes((byte)flags), data));
```

Comment:
should the old h be overwritten here? presumably the old array gets put on some GC pool, so secret state may leak? if that's the case `Clone` would need to `Array.Copy()` or something equivalent